nipype/interfaces/base/traits_extension.py

```diff
@@ -465,3 +466,100 @@ class InputMultiObject(MultiObject):
+    """Defines a new type of Tuple trait that reports inner types."""
+
+    def init_fast_validator(self, *args):
+        """Set up the C-level fast validator."""
+        super(Tuple, self).init_fast_validator(*args)
+        self.fast_validate = args
+
+    def inner_traits(self):
+        """Return the *inner trait* (or traits) for this trait."""
+        return self.types
+
+
+class Either(TraitType):
+    """Defines a trait whose value can be any of of a specified list of traits."""
+
+    def __init__(self, *traits, **metadata):
+        """Create a trait whose value can be any of of a specified list of traits."""
+        metadata['alternatives'] = tuple(trait_from(t) for t in traits)
+        self.trait_maker = _TraitMaker(
+            metadata.pop("default", None), *traits, **metadata)
+
+    def as_ctrait(self):
+        """Return a CTrait corresponding to the trait defined by this class."""
+        return self.trait_maker.as_ctrait()
+
+
+def _rebase_path(value, cwd):
+    if isinstance(value, list):
+        return [_rebase_path(v, cwd) for v in value]
+
+    try:
+        value = Path(value)
+    except TypeError:
+        pass
+    else:
+        try:
+            value = value.relative_to(cwd)
+        except ValueError:
+            pass
+    return value
+
+
+def _resolve_path(value, cwd):
+    if isinstance(value, list):
+        return [_resolve_path(v, cwd) for v in value]
+
+    try:
+        value = Path(value)
+    except TypeError:
+        pass
+    else:
+        if not value.is_absolute():
+            value = Path(cwd).absolute() / value
+    return value
+
+
+def _recurse_on_path_traits(func, thistrait, value, cwd):
+    """Run func recursively on BasePath-derived traits."""
+    if thistrait.is_trait_type(BasePath):
+        value = func(value, cwd)
+    elif thistrait.is_trait_type(traits.List):
+        innertrait, = thistrait.inner_traits
+        if not isinstance(value, (list, tuple)):
+            value = [value]
+
+        value = [_recurse_on_path_traits(func, innertrait, v, cwd)
+                 for v in value]
+    elif thistrait.is_trait_type(traits.Dict):
+        _, innertrait = thistrait.inner_traits
+        value = {k: _recurse_on_path_traits(func, innertrait, v, cwd)
+                 for k, v in value.items()}
+    elif thistrait.is_trait_type(Tuple):
+        value = tuple([_recurse_on_path_traits(func, subtrait, v, cwd)
+                       for subtrait, v in zip(thistrait.inner_traits, value)])
+    elif thistrait.alternatives:
+        is_str = [f.is_trait_type((traits.String, traits.BaseStr, traits.BaseBytes, Str))
+                  for f in thistrait.alternatives]
+        if any(is_str) and isinstance(value, (bytes, str)) and not value.startswith('/'):
+            return value
+        is_basepath = [f.is_trait_type(BasePath) for f in thistrait.alternatives]
+        if any(is_basepath):
+            subtrait = thistrait.alternatives[is_basepath.index(True)]
```

Comment:
Is this guaranteed to only have one?

Perhaps we can add:

```suggestion
            subtrait = thistrait.alternatives[is_basepath.index(True)]
            if sum(is_basepath) > 1:
                logger.debug('Multiple BasePath inner traits: %r' % thistrait.alternatives)
```

This would at least be a possible flag we could look for if something goes wrong.

Reply:
Also, what happens if one of the inner traits is a list or dict?

Reply:
Question 1: having an `Either(BasePath, BasePath)` doesn't make much sense. You can have it, but being an `Either` only one will suffice to trigger the path replacement. The only use-case I couldn't figure out is `Either(Str, BasePath)` where you cannot tell whether a value is meant to be a string or a path. I decided here `Str` is prevalent.

Question 2: I'll add a test case, something like `Either(Dict, Tuple)` and see what is happening. I bet I'm not recursing into those.

Reply:
What if I did:

```
Either(BasePath, Tuple(BasePath, Int))
```

This would be relatively common for AFNI, where you can specify a volume in a file.

Reply:
Added your use case and one more with a Dict.